java/dataset/src/main/java/org/apache/arrow/dataset/scanner/ScanOptions.java

```diff
@@ -69,4 +73,106 @@ public Optional<String[]> getColumns() {
   public long getBatchSize() {
     return batchSize;
   }
+
+  private ByteBuffer getProjection() {
+    return projection;
+  }
+
+  private ByteBuffer getFilter() {
+    return filter;
+  }
+
+  private ByteBuffer getProjectionAndFilter() {
+    return projectionAndFilter;
+  }
+
+  /**
+   * To evaluate what option was used to define Substrait Extended Expression (Project/Filter).
+   *
+   * @return Substrait Extended Expression configured for project new columns and/or apply filter
+   */
+  public ByteBuffer getSubstraitExtendedExpression() {
+    if (getProjection() != null) {
+      return getProjection();
+    } else if (getFilter() != null) {
+      return getFilter();
+    } else if (getProjectionAndFilter() != null) {
+      return getProjectionAndFilter();
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Builder for Options used during scanning.
+   */
+  public static class Builder {
+    private final long batchSize;
+    private final Optional<String[]> columns;
+    private ByteBuffer projection;
+    private ByteBuffer filter;
+    private ByteBuffer projectionAndFilter;
+
+    /**
+     * Constructor.
+     * @param batchSize Maximum row number of each returned {@link org.apache.arrow.vector.ipc.message.ArrowRecordBatch}
+     * @param columns (Optional) Projected columns. {@link Optional#empty()} for scanning all columns. Otherwise,
+     *                Only columns present in the Array will be scanned.
+     */
+    public Builder(long batchSize, Optional<String[]> columns) {
```

Comment:
Should a Builder API only enforce mandatory args in its constructor (e.g. `batchSize`)? `columns` is optional and can have its own builder method.

Reply:
Changed